### PR TITLE
[WIP] Add support for keypoint detection

### LINKED ISSUE #1361: Add OKS metric
## 🚀 Feature

Add support for keypoints inside [torchmetrics map](https://torchmetrics.readthedocs.io/en/stable/detection/mean_average_precision.html#).

### Additional context

Should be fairly simple to do that using the [iou_type](https://github.com/Lightning-AI/metrics/blob/master/src/torchmetrics/detection/mean_ap.py#L314) flag, since [torchmetrics map](https://torchmetrics.readthedocs.io/en/stable/detection/mean_average_precision.html#) is built upon [pycocotools](https://github.com/cocodataset/cocoapi/blob/master/PythonAPI/pycocotools/cocoeval.py#L110).

### PR DESCRIPTION
**Pull request by martin-kokos**
## What does this PR do?

Fixes #1361
Enabled by https://github.com/Lightning-AI/torchmetrics/pull/1928

Adds support for keypoint detection which is for example used for human pose detection. Support already is in pycocotools, so this PR just adapts the usage of that library.

- [ ] Make sure test data is suitable. (I have a test file but have to check where I got it from and check licencse)
- [ ] Test data needs to be added to `tests/_data`. Not sure how, its bundled in some zip on aws for some reason.

<details>
  <summary>Before submitting</summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
📚 Documentation preview 📚: https://torchmetrics--2967.org.readthedocs.build/en/2967/

### CODE REVIEW

File: src/torchmetrics/detection/mean_ap.py [hunk 1]
```diff
@@ -186,7 +186,7 @@ class MeanAveragePrecision(Metric):
 
         iou_type:
             Type of input (either masks or bounding-boxes) used for computing IOU. Supported IOU types are
-            ``"bbox"`` or ``"segm"`` or both as a tuple.
+            ``"bbox"``, ``"segm"``  or both as a tuple or ``"keypoints"``.
```
Comment (SkafteNicki) [inline src/torchmetrics/detection/mean_ap.py:189]: just to be sure, it should not be possible to set `iou_type=("bbox", "keypoints")`?

File: src/torchmetrics/detection/mean_ap.py [hunk 2]
```diff
@@ -373,7 +375,7 @@ class MeanAveragePrecision(Metric):
     def __init__(
         self,
         box_format: Literal["xyxy", "xywh", "cxcywh"] = "xyxy",
-        iou_type: Union[Literal["bbox", "segm"], tuple[str]] = "bbox",
+        iou_type: Union[Literal["bbox", "segm", "keypoints"], tuple[str]] = "bbox",
```
Comment (SkafteNicki) [inline src/torchmetrics/detection/mean_ap.py:378]:
```suggestion
        iou_type: Union[Literal["bbox", "segm", "keypoints"], tuple[str, str]] = "bbox",
```
just fixing the tuple type which seems to be wrong

File: src/torchmetrics/detection/helpers.py [hunk 3]
```diff
@@ -89,10 +89,10 @@ def _fix_empty_tensors(boxes: Tensor) -> Tensor:
 
 
 def _validate_iou_type_arg(
-    iou_type: Union[Literal["bbox", "segm"], tuple[str]] = "bbox",
+    iou_type: Union[Literal["bbox", "segm", "keypoints"], tuple[str]] = "bbox",
 ) -> tuple[str]:
     """Validate that iou type argument is correct."""
-    allowed_iou_types = ("segm", "bbox")
+    allowed_iou_types = ("segm", "bbox", "keypoints")
```
Comment (SkafteNicki) [inline src/torchmetrics/detection/helpers.py:95]: if we do not allow `"keypoints"` to be used together with the other iou types, we should validate for that in this function

File: src/torchmetrics/detection/helpers.py [hunk 4]
```diff
@@ -89,10 +89,10 @@ def _fix_empty_tensors(boxes: Tensor) -> Tensor:
 
 
 def _validate_iou_type_arg(
-    iou_type: Union[Literal["bbox", "segm"], tuple[str]] = "bbox",
+    iou_type: Union[Literal["bbox", "segm", "keypoints"], tuple[str]] = "bbox",
```
Comment (SkafteNicki) [inline src/torchmetrics/detection/helpers.py:92]:
```suggestion
    iou_type: Union[Literal["bbox", "segm", "keypoints"], tuple[str, str]] = "bbox",
```

File: src/torchmetrics/detection/mean_ap.py [hunk 5]
```diff
@@ -545,7 +560,7 @@ def compute(self) -> dict:
                     coco_eval.accumulate()
                     coco_eval.summarize()
                     stats = coco_eval.stats
-                    result_dict.update(self._coco_stats_to_tensor_dict(stats, prefix=prefix))
+                    result_dict.update(self._coco_stats_to_tensor_dict(stats, prefix=prefix, iou_type=self.iou_type))
```
Comment (SkafteNicki) [inline src/torchmetrics/detection/mean_ap.py:563]:
```suggestion
                    result_dict.update(self._coco_stats_to_tensor_dict(stats, prefix=prefix))
```
no need to pass the argument

File: src/torchmetrics/detection/mean_ap.py [hunk 6]
```diff
@@ -630,29 +647,43 @@ def _get_coco_datasets(self, average: Literal["macro", "micro"]) -> tuple[object
 
         return coco_preds, coco_target
 
-    def _coco_stats_to_tensor_dict(self, stats: list[float], prefix: str) -> dict[str, Tensor]:
+    def _coco_stats_to_tensor_dict(self, stats: list[float], prefix: str, iou_type: list) -> dict[str, Tensor]:
```
Comment (SkafteNicki) [inline src/torchmetrics/detection/mean_ap.py:650]:
```suggestion
    def _coco_stats_to_tensor_dict(self, stats: list[float], prefix: str) -> dict[str, Tensor]:
```

File: src/torchmetrics/detection/mean_ap.py [hunk 7]
```diff
@@ -630,29 +647,43 @@ def _get_coco_datasets(self, average: Literal["macro", "micro"]) -> tuple[object
...
         """Converts the output of COCOeval.stats to a dict of tensors."""
         mdt = self.max_detection_thresholds
+
+        if set(iou_type).intersection(("mask", "bbox")):
```
Comment (SkafteNicki) [inline src/torchmetrics/detection/mean_ap.py:654]:
```suggestion
        if set(self.iou_type).intersection(("mask", "bbox")):
```

File: src/torchmetrics/detection/mean_ap.py [hunk 8]
```diff
@@ -630,29 +647,43 @@ def _get_coco_datasets(self, average: Literal["macro", "micro"]) -> tuple[object
...
+                f"{prefix}map_small": torch.tensor([stats[3]], dtype=torch.float32),
+                f"{prefix}map_medium": torch.tensor([stats[4]], dtype=torch.float32),
+                f"{prefix}map_large": torch.tensor([stats[5]], dtype=torch.float32),
+                f"{prefix}mar_{mdt[0]}": torch.tensor([stats[6]], dtype=torch.float32),
+                f"{prefix}mar_{mdt[1]}": torch.tensor([stats[7]], dtype=torch.float32),
+                f"{prefix}mar_{mdt[2]}": torch.tensor([stats[8]], dtype=torch.float32),
+                f"{prefix}mar_small": torch.tensor([stats[9]], dtype=torch.float32),
+                f"{prefix}mar_medium": torch.tensor([stats[10]], dtype=torch.float32),
+                f"{prefix}mar_large": torch.tensor([stats[11]], dtype=torch.float32),
+            }
```
Comment (SkafteNicki) [inline src/torchmetrics/detection/mean_ap.py:668]: it should be possible to write this a bit more efficient if the only difference between the two is that `_small` has been removed for keypoint detection

File: tests/unittests/detection/__init__.py [hunk 9]
```diff
@@ -8,3 +8,7 @@
 _DETECTION_VAL = os.path.join(_PATH_ALL_TESTS, "_data", "detection", "instances_val2014_100.json")
 _DETECTION_BBOX = os.path.join(_PATH_ALL_TESTS, "_data", "detection", "instances_val2014_fakebbox100_results.json")
 _DETECTION_SEGM = os.path.join(_PATH_ALL_TESTS, "_data", "detection", "instances_val2014_fakesegm100_results.json")
+_DETECTION_OKS = os.path.join(
+    _PATH_ALL_TESTS, "_data", "detection", "person_keypoints_val2014_fakekeypoints100_results.json"
+)
+_DETECTION_OKS_VAL = os.path.join(_PATH_ALL_TESTS, "_data", "detection", "keypoint2014_val.json")
```
Comment (SkafteNicki) [inline tests/unittests/detection/__init__.py:14]: is this just the default data from pycocotools?

**Comment by Borda** [comment]
hi @martin-kokos, could you pls check the comments from Nicki? 🐿️

**Comment by SkafteNicki** [comment]
We should probably wait for the work in PR #https://github.com/Lightning-AI/torchmetrics/pull/3011 to be done. Then the refactoring of MAP is finished and we can add support for keypoint detection

File: tests/unittests/detection/__init__.py (hunk 9 above)
Comment (martin-kokos) [reply tests/unittests/detection/__init__.py:14]: Don't know where I found *2014* anymore, but in any case `person_keypoints_val2017.json` is available in the trainval2017 zip on cocodataset.org so that would need to be bundled in that file the tests fetch.

**Comment by martin-kokos** [comment]
Okay. I will revisit after #3011

**Comment by Borda** [comment]
@martin-kokos @SkafteNicki do you think we could finish it for the next release this week? 🦩

**Comment by martin-kokos** [comment]
Can't get the tests working, even pre-rebase in a reasonable time, so maybe best to redo from scratch. IMO there is too many layers of helpers and code sharing. I am not having fun and getting frustrated instead.
